### DISCUSSION TITLE
How do I use the logging feature for `acs`?

### DISCUSSION THREAD

**Discussion by partingscientist** [uBlockOrigin/uAssets#19154] [Q&A] [answered]
I was investigating `https://github.com/ABPindo/indonesianadblockrules/issues/427` and neither of these did any logging.

<details>

<summary>Attempts:</summary>

```adblock
doroni.me##+js(acs, { "target": "getComputedStyle", "needle": "", "context": "cpm", "log": true })
doroni.me##+js(acs, { "target": "getComputedStyle", "needle": "", "context": "cpm", "log": "true" })
doroni.me##+js(acs, { "target": "getComputedStyle", "needle": "", "context": "cpm", "log": "1" })
doroni.me##+js(acs, getComputedStyle, , cpm, log, true)
doroni.me##+js(acs, getComputedStyle, , cpm, log, 1)
```
</details>

Was there an undocumented change that I missed? Or an update to the wiki was forgotten?

**Comment by gorhill** [comment]
Logging works only for dev build. Are you using stable build?

**Reply by partingscientist** [reply]
I'm using dev build.

**Reply by gorhill** [reply, marked as answer]
There is a regression in the code path dealing with detecting JSON notation, which went undetected since <https://github.com/gorhill/uBlock/commit/fd036a51ee20>. The workaround until fixed is to avoid whitespace before the opening `{`:

    doroni.me##+js(acs,{ "target": "getComputedStyle", "needle": "", "context": "cpm", "log": true })

**Reply by gorhill** [reply]
Anyways, I think I will drop the JSON notation. Originally it was to be able to add optional extra parameters, but in the end using variable parameter list works better and is easier to use. However, `acs` has not been converted to use variable parameter list yet, so adding `, log, 1` does not work at the moment.

**Reply by ghajini** [reply]
so logging not available right now?

**Reply by partingscientist** [reply]
Working on my end at the moment.

While I'm at it, I have been meaning to ask this. Is there any particular reason to log both the matched script source and the matched script text when trying to break an external script? Unless I'm missing something, wouldn't the script text be always empty?

**Comment by ghajini** [comment]
on this site 
`https://ww1.1todaypk.info`
these filters should provide log
```
1todaypk.info##+js(acs, { "target": "document.createElement", "log": true })
1todaypk.info#@#+js(acs, document.documentElement, break;case $.)
```
but it doesn't

**Comment by stephenhawk8054** [comment]
Hi, anyone knows how to log `acs` now?

**Reply by ItsProfessional** [reply]
I'm pretty sure it's logged in the logger. See https://github.com/uBlockOrigin/uBlock-discussions/discussions/859#discussioncomment-8255637

**Reply by ghajini** [reply]
i think steph means logging string/regular expression matching the specified property which was previously available with `example.com##+js(acs, Property, , log, 1)`

**Reply by ItsProfessional** [reply]
I *assume* (didn't get the time to test) that it's logged in verbose mode (volume icon in the logger).

**Reply by ghajini** [reply]
ok this is how it is
![image](https://github.com/uBlockOrigin/uAssets/assets/20338483/f9afd00f-0dfc-4a27-ab0a-eed102974410)
when

```
streameast.*##+js(acs, Object.defineProperty)
```

**Reply by stephenhawk8054** [reply]
Hmm... It seems I can't see the content of the script in `Matched text` any more. Can anyone log `acs` now?

**Reply by gorhill** [reply]
Test case?

**Reply by stephenhawk8054** [reply]
STR:

1. Add `wsj.com##+js(acs, document.createElement)`
2. Use VPN to US and access `https://www.wsj.com/`

The logger just shows `Matched text` but without any contents inside.

**Reply by gorhill** [reply]
The reported `<script>` texts are really empty because those `script` tag are using `src` to load the resource. The scriptlet should ignore externally sourced `<script>` tags for when the text content is empty, I will see to change this as the current behavior is not useful.

**Reply by stephenhawk8054** [reply]
Hmm... I see. Weird that in `view-source` I see a script with the content

```html
<script data-ace-id="ace-code-11">if (window.__ace && typeof window.__ace === 'function') {
    const id = '61f2bdf7b00575078c959fee';
    const cb = () => {
      const vendorScript = document.createElement('script');
      vendorScript.src = 'https://d3h0wybtovte4j.cloudfront.net/ugeovhprh.js';
      vendorScript.setAttribute('id', 'adtoniq');
      vendorScript.async = true;
      vendorScript.addEventListener('load', function () { window.__ace('page', 'setPerfMark', ['adtoniq-loaded']);window.__ace('adtoniq', 'initListener', ['154a3e38-2228-44d6-9db6-c9707f3647ea']); });
      document.body.appendChild(vendorScript);
    }
    __ace('djcmp', 'executeOnCmpReady', [{cb, id }]);
  };</script>
```

but somehow I can't use `acs` with it.

**Reply by stephenhawk8054** [reply]
Ah ok, this works `wsj.com##+js(acs, __ace, adtoniq-loaded)`. I didn't realize that `document.createElement` is inside another local function so it actually doesn't run as inline.
